Bots permission

aman-gupta-03 opened the pull request:
The PR improves the bot permission UI tip for both `Organisation -> Bots` and `Personal -> Bots`. As reported here https://github.com/zulip/zulip/issues/24259 for user groups. Now, for `Organisation -> Bots` the tip banner is always visible to admin and only to non-admin when they do not have the required permission.

Fixes: #24155

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

1. If Administrators/moderators have the permissions to add bots -
![image](https://user-images.githubusercontent.com/81819156/215100776-9d71c124-514d-4c29-9b48-3ee003c06512.png)

2. For normal users (if they don't have the permission) - 
![image](https://user-images.githubusercontent.com/81819156/215101186-7d71fdab-8310-4a25-8ffa-10da60f821bd.png)



<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

N-Shar-ma reviewed (approved):
LGTM! Just make the 1st line of the commit message a little more specific, maybe something like "settings: Show bots permission banner only when user cannot add bots."

alya commented:
@sahil839 Do you have any additional feedback on this one? The screenshots look right, but I have not done manual testing.

sahil839 commented:
The live-update for changing the message in the banner works fine but not showing/hiding the banner. This can be done by hiding and showing the banner in `update_bot_settings_tip` itself instead of adding condition in template.

mechon03 commented:
I think I got the issue, I will try to solve this.

mechon03 commented:
@sahil839 Please review these changes.

sahil839 commented on static/js/settings_bots.js:
```diff
@@ -181,14 +181,16 @@ export function update_bot_settings_tip() {
         tip_text = $t({defaultMessage: "Only organization administrators can add generic bots."});
     } else {
-        tip_text = $t({defaultMessage: "Anyone in this organization can add bots."});
+        $(".bot-settings-tip").hide();
```
The code structure can be changed to improve readability. I think we can just return early if we do not want to show the banner after calling `hide`.

sahil839 reviewed (commented):
Posted one comment. Also, the live-update does not work correctly when the permissions are changed to allow a user to add bots when user was previously not allowed to add bots.

mechon03 commented:
@sahil839 Please review my new changes.

sov-j commented:
@mechon03 do you think it is ready for testing now?

sov-j commented:
Tested this on current Firefox (109.0). Works as expected per the spec, including live update. If there is anything else you want me to check, let me now.

sahil839 commented on static/js/settings_bots.js:
```diff
@@ -181,14 +181,19 @@ export function update_bot_settings_tip() {
+    if (page_params.is_admin) {
+        $(".bot-settings-tip").hide();
+        return;
+    } else if (page_params.realm_bot_creation_policy === permission_type.admins_only.code) {
```
There is no need of else/else-if conditions after return. The ideal structure should be a if block for checking whether the banner should be hidden and it should return after hiding the banner. And then the code to show the banner with appropriate text.

mechon03 replied:
Updated the PR with the changes, can you please take a look now. @sahil839

sahil839 reviewed (commented):
Posted one comment. Also please update the commit structure to follow the [guidelines](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html).

sahil839 commented:
Looks good to me. Would probably be better to make it clear in the commit message that banner is still shown when user can add other bots but not generic bots.

@alya Just one thing, this PR also updates the behavior of banner in `Organization > Bots` panel, the issue mentions only about `Personal > Bots` panel, so just wanted to point it out in case we do not need that change.

alya commented:
Got it, updating the org banner too sounds good.

@mechon03 Next time, please be sure to describe any deviations from the issue in the description for your PR.

alya commented:
@timabbott This PR has been reviewed by @sahil839 ; let me know if you need me to manually test it.

mechon03 commented:
Apologies, I will make sure add such deviations from next time.

aman-gupta-03 commented:
@timabbott, thanks for your feedback. I have added a new commit. Please review the changes.
Also, as reported here #24259 (task-2) for user groups I have added those changes for bots too. Permission banner for `Organisation -> Bot` should always be visible to admin but not to other users when they have the required permission.

sahil839 commented:
This does not implement Tim's suggestion and I think that's better for readability. I had to give some time to undestand the current implementation. We can just call the same functions twice for personal and organization settings panel.

Also, both the commits have exactly same commit messages, when the second commit is just a type of refactoring I guess and does not do any behavorial changes. Please fix the commit message to mention what changes have done and the reason for doing those changes.

aman-gupta-03 commented:
I will change the commit message, do you have any suggestions to what should be done differently? I found similar approach used to implement the permission banner for custom emoji. Also, this code handles cases for both `personal` and `organisational settings`. @sahil839

sahil839 commented:
I think you have mixed up changes in two commits. The changes in first commit looks like both `Organization` and `Personal` settings panels will have same banner but the commit message contradicts it. And in the second commit, you have made changes to have different behavior in `Organization` and `Personal` settings panels as mentioned in commit message.

The correct commit structure should be -
- First commit should fix the banners to behave correctly in both `Organization` and `Personal` settings panels as mentioned in the commit message using the approach Tim mentioned [here](https://github.com/zulip/zulip/pull/24198#issuecomment-1423459876).
- Second commit should be the refactoring to make the code similar to emoji settings banner to use a separate template file for the banner messages.

aman-gupta-03 commented:
Thanks for you feedback. I understand that two commits must be separated and have different commit messages.
However, I am facing problem in implementing the approach suggested by `Tim`. Some guidance there would be really helpful @sahil839. Also, my approach in the second commit works fine as per the issue, do it needs refactoring as per `Tim's` approach? Considering similar method is used to render `custom emoji tip`.

sahil839 commented:
I am ok with the changes in second commit using the new template file like custom emoji tip.

Regarding the approach suggested by Tim, it will be better for readability if we call the function twice - once for personal settings panel and once for organization settings panel. You can pass two parameters to the function - one the tip element using its id and then a bool variable `for_org_settings`. Then you can just show or hide the element and add the text according to the permissions. And the `for_org_settings` parameter should be use to decide whether we want to show the banner to admins.

timabbott commented:
This was completed via https://github.com/zulip/zulip/pull/24562.
